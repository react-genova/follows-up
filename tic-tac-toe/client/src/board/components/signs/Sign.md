### With values

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
</div>
```

### Without values (hover)

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```

### A little bigger and mixed

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"480px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```