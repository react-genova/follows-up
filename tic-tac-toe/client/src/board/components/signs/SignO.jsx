import React from 'react';
import PropTypes from 'prop-types';

const SignO = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <ellipse cx="50" cy="50" rx="46" ry="46" fill="transparent" stroke={ghost ? 'lightgray' : 'yellow'} strokeWidth="8" />
    </svg>
);

SignO.propTypes = {
    ghost: PropTypes.bool,
};

SignO.defaultProps = {
    ghost: false,
};

export default SignO;
