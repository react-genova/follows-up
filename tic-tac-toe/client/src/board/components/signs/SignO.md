### On light background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px" }}>
    <SignO />
    <SignO ghost />
</div>
```

### On dark background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```

### A little bigger

```jsx
<div style={{ display:"flex", width:"200px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```