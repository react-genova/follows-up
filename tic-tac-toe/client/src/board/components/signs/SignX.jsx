import React from 'react';
import PropTypes from 'prop-types';

const SignX = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <line x1="0" y1="0" x2="100" y2="100" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
        <line x1="0" y1="100" x2="100" y2="0" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
    </svg>
);

SignX.propTypes = {
    ghost: PropTypes.bool,
};

SignX.defaultProps = {
    ghost: false,
};


export default SignX;
