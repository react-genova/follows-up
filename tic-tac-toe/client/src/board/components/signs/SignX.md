### On light background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px" }}>
    <SignX />
    <SignX ghost />
</div>
```

### On dark background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <SignX />
    <SignX ghost />
</div>
```

### A little bigger

```jsx
<div style={{ display:"flex", width:"200px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <SignX />
    <SignX ghost />
</div>
```