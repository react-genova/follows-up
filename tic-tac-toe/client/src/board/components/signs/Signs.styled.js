import styled from 'styled-components';

const SignContainer = styled.div`
    width: 100%;
    height: 100%;
`;

export default SignContainer;
