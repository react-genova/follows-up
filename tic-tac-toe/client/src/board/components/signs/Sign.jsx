import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import * as SIGNS from '../../modules/types/signs.constants';
import SignContainer from './Signs.styled';
import SignO from './SignO';
import SignX from './SignX';

class Sign extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            hover: false,
        }
    }

    onMouseEnter = () => this.setState({ hover: true });

    onMouseLeave = () => this.setState({ hover: false });

    onClick = () => this.props.onClick(this.props.type);

    render() {
        const { type, hasValue } = this.props;
        const { hover } = this.state;
        return (
            <SignContainer onClick={this.onClick} onMouseEnter={this.onMouseEnter} onMouseLeave={this.onMouseLeave}>
            {
                (hasValue || hover) && {
                    [SIGNS.SIGN_O]: <SignO ghost={!hasValue} />,
                    [SIGNS.SIGN_X]: <SignX ghost={!hasValue} />,
                }[type]
            }
            </SignContainer>
        );
    }
}

Sign.propTypes = {
    hasValue: PropTypes.bool,
    onClick: PropTypes.func,
    type: PropTypes.oneOf([SIGNS.SIGN_O, SIGNS.SIGN_X]),
};

Sign.defaultProps = {
    hasValue: false,
    onClick: () => null,
    type: SIGNS.SIGN_O,
};

export default Sign;
