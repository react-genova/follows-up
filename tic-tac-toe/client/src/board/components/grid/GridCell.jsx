import React from 'react';
import PropTypes from 'prop-types';
import { Cell } from './Grid.styled';
import Sign from '../signs/Sign';
import { playingTypePropTypes, valuePropTypes } from './Grid.types';

const GridCell = ({
    index,
    value,
    valid,
    playingType,
    onClick,
    highlight,
}) => (
    <Cell highlight={highlight} type={value}>
        <Sign className={`cell-${index}`} hasValue={valid} type={valid ? value : playingType} onClick={() => onClick(index)} />
    </Cell>
);

GridCell.propTypes = {
    value: valuePropTypes.isRequired,
    valid: PropTypes.bool.isRequired,
    index: PropTypes.number.isRequired,
    highlight: PropTypes.bool.isRequired,
    playingType: playingTypePropTypes.isRequired,
    onClick: PropTypes.func,
};

GridCell.defaultProps = {
    onClick: undefined,
};

export default GridCell;
