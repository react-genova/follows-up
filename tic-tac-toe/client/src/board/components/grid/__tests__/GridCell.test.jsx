import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';
import GridCell from '../GridCell';
import { SIGN_O, SIGN_X } from '../../../modules/types/signs.constants';

describe('GridCell component', () => {
    afterEach(cleanup);

    it('renders without crashing', () => {
        const { container } = render(<GridCell index={0} value={SIGN_O} valid playingType={SIGN_X} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        const onClick = jest.fn();
        const INDEX = 43;
        const { container } = render(<GridCell index={INDEX} value={SIGN_O} valid playingType={SIGN_X} onClick={onClick} />);
        fireEvent.click(container.firstChild.firstChild);
        expect(onClick).toHaveBeenCalledTimes(1);
        expect(onClick).toHaveBeenCalledWith(INDEX);
    });
});
