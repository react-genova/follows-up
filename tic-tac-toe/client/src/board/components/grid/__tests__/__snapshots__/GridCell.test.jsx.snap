// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCell component renders without crashing 1`] = `
<div
  class="sc-bwzfXH kbZGzz"
  type="MOVING_SYMBOL_O"
>
  <div
    class="sc-htpNat emEaka"
  >
    <svg
      height="100%"
      viewBox="0 0 100 100"
      width="100%"
    >
      <ellipse
        cx="50"
        cy="50"
        fill="transparent"
        rx="46"
        ry="46"
        stroke="yellow"
        stroke-width="8"
      />
    </svg>
  </div>
</div>
`;
